src/meshlink.c

```diff
@@ -674,33 +522,49 @@ int check_port(meshlink_handle_t *mesh) {
 	return 0;
 }
 
-static void deltree(const char *dirname) {
-	DIR *d = opendir(dirname);
+static bool write_main_config_files(meshlink_handle_t *mesh) {
+	uint8_t buf[4096];
 
-	if(d) {
-		struct dirent *ent;
+	/* Write the main config file */
+	packmsg_output_t out = {buf, sizeof buf};
 
-		while((ent = readdir(d))) {
-			if(ent->d_name[0] == '.') {
-				continue;
-			}
+	packmsg_add_uint32(&out, MESHLINK_CONFIG_VERSION);
+	packmsg_add_str(&out, mesh->name);
+	packmsg_add_bin(&out, ecdsa_get_private_key(mesh->private_key), 96);
+	packmsg_add_bin(&out, ecdsa_get_private_key(mesh->invitation_key), 96);
+	packmsg_add_uint16(&out, atoi(mesh->myport));
 
-			char filename[PATH_MAX];
-			snprintf(filename, sizeof(filename), "%s" SLASH "%s", dirname, ent->d_name);
+	if(!packmsg_output_ok(&out)) {
+		return false;
+	}
 
-			if(unlink(filename)) {
-				deltree(filename);
-			}
-		}
+	config_t config = {buf, packmsg_output_size(&out, buf)};
 
-		closedir(d);
+	if(!main_config_write(mesh, &config)) {
+		return false;
 	}
 
-	rmdir(dirname);
+	/* Write our own host config file */
+	if(!node_write_config(mesh, mesh->self)) {
+		return false;
+	}
+
+	return true;
 }
 
-static bool finalize_join(meshlink_handle_t *mesh) {
-	char *name = xstrdup(get_value(mesh->data, "Name"));
+static bool finalize_join(meshlink_handle_t *mesh, const void *buf, uint16_t len) {
+	packmsg_input_t in = {buf, len};
+	uint32_t version = packmsg_get_uint32(&in);
+
+	if(version != MESHLINK_INVITATION_VERSION) {
+		logger(mesh, MESHLINK_ERROR, "Invalid invitation version!\n");
+		return false;
+	}
+
+	char *name = packmsg_get_str_dup(&in);
```

Comment:
Allocated `name` string is not freed after duplicating it to `mesh->self` handle parameters

Reply:
Done.